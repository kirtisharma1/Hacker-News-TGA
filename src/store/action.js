import {
  UPDATE_FEED,
  UPDATE_PAGE,
  UPDATE_UPVOTED_LIST,
  UPDATE_HIDE_LIST,
  UPDATE_SORT_TYPE,
} from "./constants";

export const updateFeed = (data, dispatch) => {
  return dispatch({
    type: UPDATE_FEED,
    payload: data,
  });
};

export const updatePage = (data, dispatch) => {
  return dispatch({
    type: UPDATE_PAGE,
    payload: data,
  });
};

export const updateUpvotedList = (data, dispatch) => {
  return dispatch({
    type: UPDATE_UPVOTED_LIST,
    payload: data,
  });
};

export const updateHideList = (data, dispatch) => {
  return dispatch({
    type: UPDATE_HIDE_LIST,
    payload: data,
  });
};

export const updateSortType = (data, dispatch) => {
  return dispatch({
    type: UPDATE_SORT_TYPE,
    payload: data,
  });
};
