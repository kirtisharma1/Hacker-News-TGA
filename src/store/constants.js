export const UPDATE_FEED = "UPDATE_FEED";
export const UPDATE_SORT_TYPE = "UPDATE_SORT_TYPE";
export const UPDATE_UPVOTED_LIST = "UPDATE_UPVOTED_LIST";
export const UPDATE_HIDE_LIST = "UPDATE_HIDE_LIST";
export const UPDATE_PAGE = "UPDATE_PAGE";
export const HEADER_TITLE = "Hacker News";
export const HEADER_TOP_LABEL = "top";
export const HEADER_NEW_LABEL = "new";
export const HIDE_LABEL = "hide";
export const MORE_LABEL = "more";
export const LOADING_TEXT = "Loading...";
