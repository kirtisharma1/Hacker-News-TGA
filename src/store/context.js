import React from "react";

export const GlobalContext = React.createContext();

export const GlobalContextProvider = ({ data, children }) => (
  <GlobalContext.Provider value={data}>{children}</GlobalContext.Provider>
);
