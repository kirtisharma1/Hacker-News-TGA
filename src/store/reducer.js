import {
  UPDATE_FEED,
  UPDATE_PAGE,
  UPDATE_UPVOTED_LIST,
  UPDATE_HIDE_LIST,
  UPDATE_SORT_TYPE,
} from "./constants";
import { HEADER_TOP_LABEL } from "../constants";

export const initialState = {
  feed: [],
  sortType: HEADER_TOP_LABEL,
  upvotedList: [],
  hideList: [],
  page: 0,
};

export const GlobalReducer = (state, action) => {
  switch (action.type) {
    case UPDATE_FEED:
      return {
        ...state,
        feed: [...action.payload],
      };
    case UPDATE_PAGE:
      return {
        ...state,
        page: action.payload,
      };
    case UPDATE_UPVOTED_LIST:
      return {
        ...state,
        upvotedList: [...action.payload],
      };
    case UPDATE_HIDE_LIST:
      return {
        ...state,
        hideList: [...action.payload],
      };
    case UPDATE_SORT_TYPE:
      return {
        ...state,
        sortType: action.payload,
      };
    default:
      return state || initialState;
  }
};
