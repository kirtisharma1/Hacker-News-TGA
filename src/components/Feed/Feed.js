import React, { useEffect, useContext } from "react";
import PropTypes from "prop-types";
import NewsItem from "../newsItem/NewsItem";
import { GlobalContext } from "../../store/context";
import { MORE_LABEL, LOADING_TEXT } from "../../constants";
import {
  updatePage,
  updateUpvotedList,
  updateHideList,
} from "../../store/action";
import useLocalStorage from "../../hooks/useLocalStorage";

export default function Feed({ isMore }) {
  const { state, dispatch } = useContext(GlobalContext);
  const { feed, page, upvotedList, hideList } = state;
  const [upvoteList, setUpvoteList] = useLocalStorage("upvotedNewsItems", []);
  const [hiddenList, setHiddenList] = useLocalStorage("hiddenNewsItems", []);

  useEffect(() => {
    updateUpvotedList(upvoteList, dispatch);
    updateHideList(hiddenList, dispatch);
    console.log(state, "state");
  }, []);

  const getType = (post) => {
    const index = upvotedList
      ? upvotedList.findIndex((item) => item.objectID === post.objectID)
      : -1;
    if (index >= 0) {
      return "\u25BC";
    }
    return "\u25B2";
  };

  const isHidden = (post) => {
    const index = hideList
      ? hideList.findIndex((item) => item.objectID === post.objectID)
      : -1;
    if (index >= 0) {
      return false;
    }
    return true;
  };

  return (
    <main className="container-wrapper feed">
      {feed.length ? (
        <>
          {feed
            .filter((post) => isHidden(post))
            .map((item, index) => (
              <NewsItem
                key={item.objectID}
                newsItem={item}
                order={index}
                isUpvoted={getType(item)}
              />
            ))}
          {isMore && (
            <button
              type="button"
              className="btn-empty feed__more"
              onClick={() => updatePage(page + 1, dispatch)}
            >
              {MORE_LABEL}
            </button>
          )}
        </>
      ) : (
        <p>{LOADING_TEXT}</p>
      )}
    </main>
  );
}

Feed.propTypes = {
  isMore: PropTypes.bool,
};

Feed.defaultProps = {
  isMore: true,
};
