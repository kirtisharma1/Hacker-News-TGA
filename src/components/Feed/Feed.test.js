import React from "react";
import { render, fireEvent, wait } from "@testing-library/react";
import Feed from "./Feed";
import { GlobalContextProvider } from "../../store/context";
import { initialState } from "../../store/reducer";
import { LOADING_TEXT, MORE_LABEL } from "../../constants";

const mockFeed = [
  {
    created_at: "2020-04-11T18:46:08.000Z",
    title: "John Conway has died",
    url: "https://twitter.com/CardColm/status/1249038195880341505",
    author: "ColinWright",
    points: 2037,
    story_text: null,
    comment_text: null,
    num_comments: 156,
    story_id: null,
    story_title: null,
    story_url: null,
    parent_id: null,
    created_at_i: 1586630768,
    _tags: ["story", "author_ColinWright", "story_22843306", "front_page"],
    objectID: "22843306",
    _highlightResult: {
      title: {
        value: "John Conway has died",
        matchLevel: "none",
        matchedWords: [],
      },
      url: {
        value: "https://twitter.com/CardColm/status/1249038195880341505",
        matchLevel: "none",
        matchedWords: [],
      },
      author: {
        value: "ColinWright",
        matchLevel: "none",
        matchedWords: [],
      },
    },
  },
  {
    created_at: "2020-04-12T02:22:08.000Z",
    title: "Nasty macOS flaw is bricking MacBooks: Don't install this update",
    url:
      "https://www.tomsguide.com/news/nasty-macos-flaw-is-bricking-macbooks-dont-install-this-update",
    author: "Corrado",
    points: 531,
    story_text: null,
    comment_text: null,
    num_comments: 355,
    story_id: null,
    story_title: null,
    story_url: null,
    parent_id: null,
    created_at_i: 1586658128,
    _tags: ["story", "author_Corrado", "story_22845629", "front_page"],
    objectID: "22845629",
    _highlightResult: {
      title: {
        value:
          "Nasty macOS flaw is bricking MacBooks: Don't install this update",
        matchLevel: "none",
        matchedWords: [],
      },
      url: {
        value:
          "https://www.tomsguide.com/news/nasty-macos-flaw-is-bricking-macbooks-dont-install-this-update",
        matchLevel: "none",
        matchedWords: [],
      },
      author: { value: "Corrado", matchLevel: "none", matchedWords: [] },
    },
  },
];

const mockItem = {
  created_at: "2020-04-11T18:46:08.000Z",
  title: "John Conway has died",
  url: "https://twitter.com/CardColm/status/1249038195880341505",
  author: "ColinWright",
  points: 2037,
  story_text: null,
  comment_text: null,
  num_comments: 156,
  story_id: null,
  story_title: null,
  story_url: null,
  parent_id: null,
  created_at_i: 1586630768,
  _tags: ["story", "author_ColinWright", "story_22843306", "front_page"],
  objectID: "22843306",
  _highlightResult: {
    title: {
      value: "John Conway has died",
      matchLevel: "none",
      matchedWords: [],
    },
    url: {
      value: "https://twitter.com/CardColm/status/1249038195880341505",
      matchLevel: "none",
      matchedWords: [],
    },
    author: {
      value: "ColinWright",
      matchLevel: "none",
      matchedWords: [],
    },
  },
};

const dispatch = jest.fn();
describe("Feed component", () => {
  let result;
  beforeEach(() => {
    result = render(
      <GlobalContextProvider
        data={{ state: { ...initialState, feed: mockFeed }, dispatch }}
      >
        <Feed isMore={true} />
      </GlobalContextProvider>
    );
  });
  test("snapshot", () => {
    expect(result.asFragment()).toMatchSnapshot();
  });

  test("loading ", () => {
    result.rerender(
      <GlobalContextProvider data={{ state: initialState, dispatch }}>
        <Feed isMore={true} />
      </GlobalContextProvider>
    );
    expect(result.getByText(LOADING_TEXT)).toBeInTheDocument();
  });

  test("upvoted list ", () => {
    result.rerender(
      <GlobalContextProvider
        data={{
          state: { ...initialState, feed: mockFeed, upvotedList: [mockItem] },
          dispatch,
        }}
      >
        <Feed isMore={true} />
      </GlobalContextProvider>
    );
    expect(result.getAllByText("\u25BC").length).toBe(1);
    expect(result.getAllByText("\u25B2").length).toBe(1);
  });

  test("hidden list ", () => {
    result.rerender(
      <GlobalContextProvider
        data={{
          state: { ...initialState, feed: mockFeed, hideList: [mockItem] },
          dispatch,
        }}
      >
        <Feed isMore={true} />
      </GlobalContextProvider>
    );
    const button = result.getByText(MORE_LABEL);
    fireEvent.click(button);
    expect(result.queryAllByText("John Conway has died").length).toBe(0);
  });
});
