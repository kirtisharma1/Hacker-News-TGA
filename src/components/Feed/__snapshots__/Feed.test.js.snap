// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Feed component snapshot 1`] = `
<DocumentFragment>
  <main
    class="container-wrapper feed"
  >
    <article
      class="news-item  even"
      data-testid="news-item"
    >
      <span
        class="news-item__comment-count"
      >
        156
      </span>
      <span
        class="news-item__points news-item--lightest"
      >
        2037
      </span>
      <button
        class="btn-empty news-item__upvote"
        type="button"
      >
        ▲
      </button>
      <p
        class="news-item__title"
      >
        John Conway has died
        <span>
          (
          <a
            href="https://twitter.com/CardColm/status/1249038195880341505"
            rel="noopener noreferrer"
            target="_blank"
          >
            twitter.com
          </a>
          ) by 
          <b>
            ColinWright
          </b>
           
          <date>
            1 day ago
          </date>
        </span>
         [
        <button
          class="btn-empty news-item__hide"
          type="button"
        >
          hide
        </button>
        ]
      </p>
    </article>
    <article
      class="news-item  odd"
      data-testid="news-item"
    >
      <span
        class="news-item__comment-count"
      >
        355
      </span>
      <span
        class="news-item__points news-item--lightest"
      >
        531
      </span>
      <button
        class="btn-empty news-item__upvote"
        type="button"
      >
        ▲
      </button>
      <p
        class="news-item__title"
      >
        Nasty macOS flaw is bricking MacBooks: Don't install this update
        <span>
          (
          <a
            href="https://www.tomsguide.com/news/nasty-macos-flaw-is-bricking-macbooks-dont-install-this-update"
            rel="noopener noreferrer"
            target="_blank"
          >
            www.tomsguide.com
          </a>
          ) by 
          <b>
            Corrado
          </b>
           
          <date>
            1 day ago
          </date>
        </span>
         [
        <button
          class="btn-empty news-item__hide"
          type="button"
        >
          hide
        </button>
        ]
      </p>
    </article>
    <button
      class="btn-empty feed__more"
      type="button"
    >
      more
    </button>
  </main>
</DocumentFragment>
`;
